File: rust/src/utils/image.rs
use std::fmt::Display;

use crate::utils::resolution::Resolution;

use super::pixel::Pixel;

#[derive(Debug, Clone, Copy)]
pub enum ImageType {
    PackedImage,
    StrideImage
}

#[derive(Debug, PartialEq)]
pub enum Image {
    PackedImage(PackedImage),
    StrideImage(StrideImage)
}

#[derive(Debug, PartialEq)]
pub struct PackedImage {
    resolution: Resolution,
    pixels: Vec<Pixel>
}

impl PackedImage {
    pub fn new(resolution: Resolution, pixels: Vec<Pixel>) -> Self { 
        Self { resolution, pixels } 
    }
}

impl Display for PackedImage {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}\n{:?}", self.resolution, self.pixels)
    }
}

#[derive(Debug, PartialEq)]
pub struct StrideImage {
    pub resolution: Resolution,
    pub pixels_red: Vec<u8>,
    pub pixels_green: Vec<u8>,
    pub pixels_blue: Vec<u8>,
    pub pixels_alpha: Vec<u8>,
}

impl StrideImage {
    pub fn new(resolution: Resolution, pixels: Vec<Pixel>) -> Self {
        let mut res = Self {
            resolution,
            pixels_red: Vec::new(),
            pixels_green: Vec::new(),
            pixels_blue: Vec::new(),
            pixels_alpha: Vec::new(),
        };
        res.split_pixel_components(pixels);
        res
    }

    fn split_pixel_components(&mut self, pixels: Vec<Pixel>) {
        pixels.into_iter().for_each(|p| {
            self.pixels_red.push(p.red);
            self.pixels_green.push(p.green);
            self.pixels_blue.push(p.blue);
            self.pixels_alpha.push(p.alpha);
        })
    }

    pub fn merge_pixel_components(&self) -> Vec<Pixel> {
        self.pixels_red.iter()
            .zip(self.pixels_green.iter())
            .zip(self.pixels_blue.iter())
            .zip(self.pixels_alpha.iter())
            .map(|(((&r, &g), &b), &a)| {
                Pixel::new(r, g, b, a)
            }).collect()
    }
}


impl Into<StrideImage> for PackedImage {
    fn into(self) -> StrideImage {
        StrideImage::new(self.resolution, self.pixels)
    }
}

impl Into<PackedImage> for StrideImage {
    fn into(self) -> PackedImage {
        PackedImage::new(self.resolution, self.merge_pixel_components())
    }
}

File: python/utils/image.py
#!/usr/bin/env python3

from typing import List
from utils.resolution import Resolution
from enum import Enum
import numpy as np


class ImageType(Enum):
    PackedImageType = 0
    StrideImageType = 1

class PackedImage:
    def __init__(self, resolution: Resolution, pixels: np.ndarray):
        self.resolution: Resolution = resolution
        self.pixels: np.ndarray = pixels
    
    def __str__ (self):
        return str(self.resolution) + '\n' + ' '.join(str(p) for p in self.pixels[:, :])

    def __eq__(self, other):
        return self.resolution == other.resolution and np.all(self.pixels == other.pixels)

class StrideImage:
    def __init__(self, resolution: Resolution, pixels: np.ndarray):
        self.resolution: Resolution = resolution
        self.pixels_red: np.ndarray = np.zeros((resolution.height, resolution.width))
        self.pixels_green: np.ndarray = np.zeros((resolution.height, resolution.width))
        self.pixels_blue: np.ndarray = np.zeros((resolution.height, resolution.width))
        self.pixels_alpha: np.ndarray = np.zeros((resolution.height, resolution.width))
        
        self.__split_pixel_components(pixels)

    def merge_pixel_components(self) -> np.ndarray:
        pixels = np.zeros((self.resolution.height, self.resolution.width, 4))
        pixels[:, :, 0] = self.pixels_red
        pixels[:, :, 1] = self.pixels_green
        pixels[:, :, 2] = self.pixels_blue
        pixels[:, :, 3] = self.pixels_alpha
        return pixels
    
    def __split_pixel_components(self, pixels: np.ndarray):
        self.pixels_red   = pixels[:, :, 0]
        self.pixels_green = pixels[:, :, 1]
        self.pixels_blue  = pixels[:, :, 2]
        self.pixels_alpha = pixels[:, :, 3]
    
    def __str__ (self):
        return str(self.resolution) + '\n' + \
            ' '.join(str(p) for p in self.pixels_red[:, :]) + \
            ' '.join(str(p) for p in self.pixels_green[:, :]) + \
            ' '.join(str(p) for p in self.pixels_blue[:, :]) + \
            ' '.join(str(p) for p in self.pixels_alpha[:, :])

    def __eq__(self, other):
        return self.resolution == other.resolution and \
            np.all(self.pixels_red == other.pixels_red) and \
            np.all(self.pixels_green == other.pixels_green) and \
            np.all(self.pixels_blue == other.pixels_blue) and \
            np.all(self.pixels_alpha == other.pixels_alpha)

def to_stride_image(image: PackedImage) -> StrideImage:
    return StrideImage(image.resolution, image.pixels)

def to_image(stride_image: StrideImage) -> PackedImage:
    return PackedImage(stride_image.resolution, stride_image.merge_pixel_components())